pkg/innerring/blockchain.go

```diff
@@ -0,0 +1,707 @@
+				x.log.Info("stopping distillation of blockchain notary requests (channel of mempool events is closed)")
+				return
+			}
+
+			if rawEvent.Type != mempoolevent.TransactionAdded {
+				x.log.Debug("ignore mempool event by type", zap.Stringer("type", rawEvent.Type))
+				continue
+			}
+
+			rawNotaryRequest, ok := rawEvent.Data.(*payload.P2PNotaryRequest)
+			if !ok {
+				x.log.Debug("ignore mempool event by payload type", zap.String("type", fmt.Sprintf("%T", rawEvent.Data)))
+				continue
+			}
+
+			if rawNotaryRequest.MainTransaction == nil {
+				x.log.Debug("ignore notary request (nil main transaction)")
+				continue
+			}
+
+			if !rawNotaryRequest.MainTransaction.HasSigner(x.contracts.get(contractProxy)) {
+				x.log.Debug("ignore notary request (missing Proxy contact signer)")
+				continue
+			}
+
+			ev, err := preparer.Prepare(rawNotaryRequest)
+			if err != nil {
+				switch {
+				case errors.Is(err, event.ErrTXAlreadyHandled):
+					x.log.Debug("ignore notary request (main transaction has already been handled)")
+				case errors.Is(err, event.ErrMainTXExpired):
+					x.log.Debug("ignore notary request (main transaction is expired)")
+				default:
+					x.log.Error("ignore notary request (validation failed)", zap.Error(err))
+				}
+
+				continue
+			}
+
+			contract, ok := x.contracts.resolve(ev.ScriptHash())
+			if !ok {
+				x.log.Info("ignore notary request from unknown contract: %w", zap.Stringer("address", ev.ScriptHash()))
+				continue
+			}
+
+			chDst <- notaryRequest{
+				mainTx:   *rawNotaryRequest.MainTransaction,
+				contract: contract,
+				method:   ev.Type().String(),
+				ops:      ev.Params(),
+			}
+		}
+	}
+}
+
+// Returns keys.PublicKeys of the NeoFS Sidechain committee as NeoFS Alphabet
+// members.
+func (x *blockChain) alphabetMembers() (keys.PublicKeys, error) {
+	// TODO: consider caching
+	cmt, err := x.sidechain.Committee()
+	if err != nil {
+		return nil, fmt.Errorf("get committee list from the blockchain: %w", err)
+	}
+
+	return cmt, nil
+}
+
+// alphabetLetterForAccount returns GlagoliticLetter of the NeoFS Alphabet
+// member for the specified wallet.Account. Returns negative value without an
+// error if the account is not bound to any Alphabet member.
+func (x *blockChain) alphabetLetterForAccount(acc *wallet.Account) (GlagoliticLetter, error) {
+	key := acc.PublicKey()
+	if key != nil {
+		alphabetKeys, err := x.alphabetMembers()
+		if err != nil {
+			return 0, fmt.Errorf("form Alphabet list: %w", err)
+		}
+
+		for i := range alphabetKeys {
+			if alphabetKeys[i].Equal(key) {
+				return GlagoliticLetter(i), nil
+			}
+		}
+	}
+
+	return -1, nil
+}
+
+// alphabetContract returns address of the NeoFS Alphabet contract associated
+// with the given GlagoliticLetter.
+func (x *blockChain) alphabetContract(letter GlagoliticLetter) (res util.Uint160, err error) {
+	res = getAlphabetContract(x.contracts, letter)
+	if res == (util.Uint160{}) {
+		// TODO: request from chain again?
```

Comment:
It's not supposed to happen unless you're requesting some uninitialized letter, but it's not initialized because it's not used and if it's not used then why is it requested?